DRAFT [usb_aon] Deassert pullups if SENSE lost when active

alees24 opened the pull request:
This (resurrected) PR includes two proposed changes to the 'usbdev_aon_wake' module:

- Deassert any active pullups if VBUS/SENSE is lost through disconnection when AON/Wakeup module is active.
Avoid potential confusion between host and device, and give the software longer to reinitialize when waking. Related to #18562.

- Introduce explicit reporting of the 'bus not idle' event, which previously had to be inferred by the absence of any other reported reason for waking from Deep Sleep.

Regarding the sequence of operations when resuming from Deep Sleep, software should surely consult the 'wake_events' register before deciding what to do. If a Disconnect has been detected and reported, then the bus and open pipes must be regarded as being in an indeterminate state, and it seems preferable for the aon_wake module to deassert the pull ups to prevent automatic reconnection and then leave software to decide how to resume communications upon return from sleep.

Both changes give software greater visibility into events occurring on the bus and better error handling/recovery in the event that bus signaling does not proceed as expected. For example, software presently has no way to detect the USB becoming non-idle during a transition to a sleep state. Whilst the USB protocol specification requires that resume signaling have a minimum duration long enough for the chip to sleep and resume from sleep, relying upon the correct USB behavior being observed seems undesirable.

@a-will

alees24 committed:
[usb_aon] Deassert pullups if SENSE lost when active

Deassert any active pullups if VBUS/SENSE is lost through
disconnection when AON/Wakeup module is active.
Avoid potential confusion between host and device,
and give the software longer to reinitialize when waking.

Signed-off-by: Adrian Lees <a.lees@lowrisc.org>

a-will commented:
These are just preliminary thoughts, as I haven't had time to think through all of this fully.

If the AON module can control the pull-up and respond to events, we may end up with some awkward scenarios.

Currently, we have software turn on full-power usbdev's pull-up before reading events that the AON module captured. If the AON module has removed the pull-up, then usbdev can proceed erroneously to an active state and repeatedly think it has been given a bus reset.

If we flip the order and first read the AON wake events, software may miss events that happen between the CSR read and bringing usbdev online. usbdev can no longer be in a state where it is monitoring the link at the same time as the AON module--That overlap helped ensure a clean hand-off.

alees24 committed:
[usb,aon_wake] Report 'bus not idle' status to usbdev

The 'bus not idle' condition had to be inferred by software,
from the absence of Bus Reset and Disconnect indications;
less than ideal from a robustness perspective even when returning
from Deep Sleep, but it afforded no opportunity to indicate a
non-idle bus state whilst still suspending/entering sleep.
Introduce an explicit indication for increased robustness and
error recovery.

Signed-off-by: Adrian Lees <a.lees@lowrisc.org>

a-will commented on hw/ip/pinmux/data/pinmux.hjson.tpl:
```diff
@@ -302,7 +313,7 @@
       package: "",
       desc:    '''
-               Event signal that indicates what happened while monitoring.
+               Event signal that indicates what USB SENSE was lost while monitoring.
```
Super nitty, minor typo, heh:

```suggestion
               Event signal that indicates that USB SENSE was lost while monitoring.
```

a-will reviewed (commented):
Are there any issues if we lose VBUS between reading `wake_events` and enabling the pull-up in full-power usbdev? Or... what is the order of operations that can prevent events from being lost?

Previously, the AON module couldn't disable pull-ups, so we could start up full-power usbdev and have it monitor the bus with an overlap. This would allow software to read `wake_events` and disable the AON module; the full-power usbdev module would pick up anything after reading that CSR.

If we _don't_ go straight to usbdev reaching the pull-up on state first, then it can't fully do that (at least with the state machine as it is now). If the AON module needs to be able to de-assert the pull-up at any time, do we need the full-power usbdev module to do something different (capture events a different way) for a clean hand-off?

a-will commented:
Maybe we could still tell full-power usbdev to connect before releasing the AON module (so usbdev_linkstate is still observing events), but the AON module's `sense_lost` could gate `usb_pwr_sense` until the AON module is disabled? Then usbdev_linkstate's state machine wouldn't proceed into incorrect states. Software will not be able to immediately turn on the pull-up if VBUS was lost under the AON module's watch.

I'm not sure if we would need it, but `sense_lost` could potentially mix into the `event_disconnect` INTR_STATE trigger as well (again, while the AON module is active), if we also needed software to be notified if VBUS just happens to go away after we read `wake_events`.

Would that work?

rswarbrick commented:
Hi @alees24! Is this PR still current? If so, is it worth rebasing or should it be closed?
